packages/plugin-ext/src/main/node/plugin-deployer-impl.ts

```diff
@@ -131,8 +131,8 @@ export class PluginDeployerImpl implements PluginDeployer {
 
             for (const deployerEntry of pluginDeployerEntries) {
                 const metadata = await this.pluginDeployerHandler.getPluginMetadata(deployerEntry);
-                if (metadata && !pluginsToDeploy.has(metadata.model.id)) {
-                    pluginsToDeploy.set(metadata.model.id, deployerEntry);
+                if (metadata && !pluginsToDeploy.has(metadata.model.uniqueId)) {
```

Comment:
`metadata.model.id` is already unique, the check is already in place

Reply:
for `metadata.model.extensionDependencies` will add a prefix of `vscode:extension/`
see https://github.com/theia-ide/theia/blob/85b75768f702656db4ee8ff6eaad303be8102643/packages/plugin-ext-vscode/src/node/scanner-vscode.ts#L66

Reply:
But at the end the extension dependency will be resolved to `publisher.name` and should be filtered out by `pluginsToDeploy.has(metadata.model.id)`, since there is already an extension with such publisher and name installed from local vsix.

`metadata.model.id` has to be unique identifier among all extensions, there should not be another

Reply:
Sorry I misunderstood your question, the reason I use metadata.model.uniqueId is that:
if use `pluginsToDeploy.set(metadata.model.id, deployerEntry)`, considering `queue.push(...metadata.model.extensionDependencies)`, 
then `queue` will have some extensionDependencies elements like '['vscode:extension/redhat.java']',
then `if (visited.has(current) || pluginsToDeploy.has(current)) {` will not be true.
Because the keys of `pluginsToDeploy` and the elements of `queue` will never be equal even for the extensionDependencies.

Reply:
It is expected.

Reply:
But we should continue to resolve the extensionDependencies only when pluginsToDeploy doesn't contain them.

Reply:
Why not?

Reply:
Because we will resove twice for the same plugin.